WebApplication/Processing/Publisher.cs

```diff
@@ -45,30 +45,54 @@ public Publisher(DesignAutomationClient client, ILogger<Publisher> logger, Resou
             _postProcessing = postProcessing;
         }
 
-        public async Task<WorkItemStatus> RunWorkItemAsync(Dictionary<string, IArgument> workItemArgs, ForgeAppBase config)
+        public async Task<WorkItemStatus> RunWorkItemAsync(Dictionary<string, IArgument> workItemArgs, 
+            ForgeAppBase config, string callbackUrl)
         {
+
+            // Use callback if the URL was provided
+            bool useCallback = !String.IsNullOrEmpty(callbackUrl);
+            if (useCallback)
+            {
+                XrefTreeArgument callbackOnComplete = new XrefTreeArgument()
+                {
+                    Verb = Verb.Post,
+                    Url = callbackUrl
+                };
+
+                workItemArgs.Add("onComplete", callbackOnComplete);
+            }
+
             // create work item
             var wi = new WorkItem
             {
                 ActivityId = await GetFullActivityId(config),
                 Arguments = workItemArgs
             };
 
-            // run WI and wait for completion
-            System.Diagnostics.Stopwatch sw = System.Diagnostics.Stopwatch.StartNew();
-            WorkItemStatus status = await _client.CreateWorkItemAsync(wi);
-            Trace($"Created WI {status.Id}");
-            while (status.Status == Status.Pending || status.Status == Status.Inprogress)
+            // Fire and forget vs wait for results
+            if (useCallback)
             {
-                await Task.Delay(2000);
-                status = await _client.GetWorkitemStatusAsync(status.Id);
+                await _client.CreateWorkItemAsync(wi);
```

Comment:
Creation of WI might be failed.  Are you sure the callback will be called if WI failed to create?

Plus - it would be good to trace id of the created WI.

A thought - can WI id be used instead of ClientId?

Reply:
Say - what if we use something like https://github.com/Autodesk-Forge/forge-configurator-inventor/blob/master/WebApplication/State/Uploads.cs to keep mapping between WI ID and all required info? 
Maybe it would be enough to have WI id in callback URL?  So we can generate it on-fly.  And no need to pass it.

Reply:
Good point about the failed WI creation. I expected exception to be thrown in that case that would bubble all the way up and get eventually processed in the JobHub just like polling version of the workflow. Wouldn't that be the case?

Reply:
Regarding holding the state instead of passing it back in the callback. So the idea is to just pass ID in the callback and then map all the info using cached structure? When we discussed similar idea with @mivasi the conclusion for now was that it would be nice if we can avoid any resource sharing and make the callback fully discrete. But maybe what you are saying is better idea :)

Reply:
> Good point about the failed WI creation. I expected exception to be thrown in that case that would bubble all the way up and get eventually processed in the JobHub just like polling version of the workflow. Wouldn't that be the case?

@mivasi Do you remember the details here?  IIRC - it can return something like "bad instructions" in some cases, but not 100% sure.

Reply:
I will create new task to specifically look at that, thank you Eugene :)